examples/states_ui/pause.rs

```diff
@@ -0,0 +1,107 @@
+const RESUME_BUTTON_ID: &str = "resume";
+const EXIT_TO_MAIN_MENU_BUTTON_ID: &str = "exit_to_main_menu";
+const EXIT_BUTTON_ID: &str = "exit";
+
+// load the pause_menu.ron prefab then instantiate it
+// if the "resume" button is clicked, goto MainGameState
+// if the "exit_to_main_menu" button is clicked, remove the pause and main game states and go to MenuState.
+// if the "exit" button is clicked, quit the program.
+impl<'a> SimpleState for PauseMenuState {
+    fn on_start(&mut self, data: StateData<GameData>) {
+        let world = data.world;
+
+        self.root =
+            Some(world.exec(|mut creator: UiCreator<'_>| creator.create("ui/pause_menu.ron", ())));
+    }
+
+    fn on_stop(&mut self, data: StateData<GameData>) {
+        if let Some(root) = self.root {
+            if data.world.delete_entity(root).is_ok() {
+                self.root = None;
+            }
+        }
+        self.resume_button = None;
+        self.exit_to_main_menu_button = None;
+    }
+
+    fn handle_event(&mut self, data: StateData<GameData>, event: StateEvent) -> SimpleTrans {
+        match event {
+            StateEvent::Window(event) => {
+                if is_close_requested(&event) || is_key_down(&event, VirtualKeyCode::Escape) {
+                    Trans::Pop
+                } else {
+                    Trans::None
+                }
+            }
+            StateEvent::Ui(UiEvent {
+                event_type: UiEventType::Click,
+                target,
+            }) => {
+                if Some(target) == self.resume_button {
+                    log::info!("Resuming Game!");
+                    Trans::Pop
+                } else if Some(target) == self.exit_to_main_menu_button {
+                    let mut state_transition_event_channel = data
+                        .world
+                        .write_resource::<EventChannel<TransEvent<GameData, StateEvent>>>();
+
+                    log::info!("Exiting Game, Switching to MainMenu!");
+
+                    // this allows us to first 'Pop' this state, and then exchange whatever was
+                    // below that with a new MainMenu state.
+                    state_transition_event_channel.single_write(Box::new(|| Trans::Pop));
+                    state_transition_event_channel
+                        .single_write(Box::new(|| Trans::Switch(Box::new(MainMenu::default()))));
+
+                    Trans::None // we could also not add the pop to the channel and Pop here
+                                // but like this the execution order is guaranteed (in the next versions)
+                } else if Some(target) == self.exit_button {
+                    Trans::Quit
+                } else {
+                    Trans::None
+                }
+            }
+            _ => Trans::None,
+        }
+    }
+
+    fn update(&mut self, data: &mut StateData<GameData>) -> SimpleTrans {
+        data.data.update(&data.world);
```

Comment:
Given `PauseMenuState` implements `SimpleState`, this is done already in the [blanket implementation](https://github.com/amethyst/amethyst/blob/03cac16f94bf569f9c99d5c71db25c43951fe7ff/src/state.rs#L336), so this can be removed, otherwise it would be double-updating in the same tick.

Reply:
Ah, I see. Yes, I will change that then.